### ISSUE TITLE
Proposal: Move non-JS-specific optimizations to the CoreFn optimization pass

### ISSUE THREAD

**Issue by PureFunctor** [purescript/purescript#4206]
## Proposal
Certain optimizations such as inlining `$`/`#` is currently performed in the `L.PS.CoreImp.Optimizer` module as part of JS code generation. The proposal is that these should be moved into `L.PS.CoreFn.Optimizer` to allow downstream consumers of the `CoreFn` API to benefit from non-JS-specific optimizations.

## Motivation

The discussion originally started in #4205 where I implemented further inlining of `apply` by catching it during JS code generation. But if we ought to ensure that we keep all our optimizations in one place, we run into some potential problems. For one, in the PR, moving that optimization over to the `CoreImp` inliner makes it harder to distinguish how to inline newtype applications as the `AST` type doesn't have the same metadata that `CoreFn` does. In hindsight, after implementing the original PR, I realized that this could've been a likely fit for `CoreFn` instead but considering that said code path is relatively new, I was hesitant to add changes.

The primary motivation for this is to make way for #2866/#4205, and eventually a more general optimization framework that separates itself from the JS backend. I'd be glad to hear thoughts from backend maintainers on this as well.

**Comment by nwolverson** [comment]
I think one needs to be a little careful with the definition of "JS-specific" here. There are optimisations that are certainly JS-specific (eg tied to the specifics of the generated code & language/platform features); but there are also optimisations that are done "for" JS (in the sense that JS performance justifies them) that could be done in CoreFn, I question if they should be.

In particular, there are currently 2 corefn optimisations in place, "unused partial fn" and ["closed record update"](https://github.com/purescript/purescript/blob/master/src/Language/PureScript/CoreFn/Optimizer.hs#L30). The latter transforms record updates to record literals, which is a JS performance gain, but a performance loss for another backend (`purerl`) and crucially a loss of information.

I wonder if the latter category of optimsation should be performed at the corefn level, if that is the best place to do so, but post CoreFn emit.

**Comment by PureFunctor** [comment]
Thank you for the insights. When I originally wrote this proposal, I intended "JS-specific" to mean anything that the "canonical" JavaScript code generator for PureScript uses, whether that'd be because of platform features or for JS performance gains like you mentioned which directly impacts downstream. I'm not entirely sure what the stance/general agreement is on whether the CoreFn API and its consumers (alternative backends, external optimizers) should be treated as public and first-class, but I think moving to this direction would ultimately be beneficial overall.

I think that the first step to going about this is to separate the CoreFn optimizations into "public" and "private" modes: public pertains to anything that gets emitted to the corefn.json file or directly consumed by external backends, while private pertains to all other optimizations pre-`L.PS.CodeGen.JS.moduleToJs`. I haven't extensively walked through the compilation pipeline, but I do believe that this should be possible as JS codegen is called much later than the initial CoreFn optimization stage.

With this in mind, I do think I should expand this proposal to address the points above, namely how we expose the CoreFn API.

**Comment by f-f** [comment]
> The latter transforms record updates to record literals, which is a JS performance gain, but a performance loss for another backend

I was also wondering about this. It might be worth exposing some flags to turn some CoreFn optimizations on/off, so that it'd be possible to exclude some when necessary?

**Comment by JordanMartinez** [comment]
> It might be worth to expose some flags to turn some CoreFn optimizations on/off, so that it'd be possible for backends to exclude some when necessary?

To me, this sort of feels like 'language extensions' but for backends.

**Comment by f-f** [comment]
> To me, this sort of feels like 'language extensions' but for backends

Could you elaborate more on this?

The main issue I see with language extensions (as they are in Haskell, I assume that's the model we're talking about) is the fact that certain combinations are not compatible with others, leading to an explosion of the compatibility matrix (which means more complexity in the codebase, etc).

"Toggling optimizations" is a very different beast: optimizations are, by definition, semantics-preserving, and your code should work without them, with all of them applied, and with anything in between - if that's not the case, it's a bug! With extensions it's sort of expected that your code will work differently, as they can affect the semantics.

If your observation above is about the codebase complexity brought by extensions I'm not sure I see it, as optimizations in CoreFn are just applied sequentially, and toggling some off is just about removing things from this list:
https://github.com/purescript/purescript/blob/18e3b14c556c5bd58e391531e4f4e511f834c3ff/src/Language/PureScript/CoreFn/Optimizer.hs#L27

**Comment by rhendric** [comment]
> your code should work without them, with all of them applied, and with anything in between - if that's not the case, it's a bug!

That to me implies that we should have tests that ensure that your code will work without them, with all of them applied, and with anything in between. So, what, should we run the entire test suite 2<sup>n</sup> times? Or should we maintain a matrix of which tests should be rerun with which combinations of optimizations? It may not be as bad as maintaining compilers for 2<sup>n</sup>-ish separate languages, as is the case with GHC, but it is still like maintaining 2<sup>n</sup>-ish compilers for the same language, which I don't think is much better.

On a separate note, I have come to the conclusion that `optimizeUnusedPartialFn` is garbage and I have a PR in the wings to remove it. So for now, we're only talking about `optimizeClosedRecordUpdate`, which should just be moved closer to, if not into, the JS backend code, and #4205, which... can anyone think of a reason to want to turn that off?

**Comment by f-f** [comment]
> So, what, should we run the entire test suite 2n times? Or should we maintain a matrix of which tests should be rerun with which combinations of optimizations?

I don't think this implies the need for any additional testing, if we'd be worried about optimisations being correct we should be testing them in isolation anyways!
I'm also wondering about the 2^n number there, I think it's more like O(n^2) cases, since we wouldn't change the order in which they are applied.
In any case, I feel that there's no need to write these by hand and that a Quickcheck test flipping flags should be good to have enough assurance here.

**Comment by JordanMartinez** [comment]
> The main issue I see with language extensions (as they are in Haskell, I assume that's the model we're talking about) is the fact that certain combinations are not compatible with others, leading to an explosion of the compatibility matrix (which means more complexity in the codebase, etc).

That is one issue, but I think the real one is that one ultimately has multiple languages as a result rather than just one (basically, what @rhendric said above). 

I could be misunderstanding a lot here in this paragraph, but I'll say still say it. Assuming we decouple the JS backend from this compiler at some point, the language that remains is `CoreFn`. I think there should only ever be one `CoreFn` language. As for what each backend does with that language, that's up to them. In some ways, I think the current issue is a side-effect of the reality that the JS backend is still tightly coupled to the compiler.

**Comment by rhendric** [comment]
Any objections here to merging #4205? This proposal is a good thing to discuss but I don't think resolving it in the general case should necessarily hold up the specific case of #4205. We've had some time to mull it over and I haven't heard any potential risks to that particular change.

(I intend to click the merge button in 48 hours if no response.)

**Comment by PureFunctor** [comment]
Are there any backends that depend on function composition to be explicitly present in the CoreFn for performance or metadata? I have a partial implementation for this that is essentially a port from CoreImp, though I'm not sure how this affects downstream.
